### DISCUSSION TITLE
SinglePhaseFluidProperties & AD system

### DISCUSSION THREAD

**Discussion by joe61vette** [idaholab/moose#23600] [Q&A General] [answered]
Hi:

I am writing a very simple fluid property but having trouble getting the AD system to produce the derivatives.  I inherit from SinglePhaseFluidProperties.  The function is:

`ADReal
ConstantFluidProperties::h_from_p_T(const ADReal & /*p*/, const ADReal & T) const
{
  return _cp * (T - 273.15);
}`

where _cp is a constant.

In my FluidMaterial::computeQpProperties(), I have:

`_h[_qp] = _fluid.h_from_p_T(_p[_qp], _T[_qp]);`

where in the constructor initialization:

_p(adCoupledValue(CRAB::pressure)),
_T(adCoupledValue(CRAB::T_fluid)),
_h(declareADProperty<Real>(CRAB::enthalpy)),

When I print the values of _h[_qp] and _T[_qp], it shows no derivs:   T = (293.15,{})  h = (83682,{})

In the test problem, T is a non-linear variable.  However, in the FluidMaterial validParams, I have:

`params.addCoupledVar(CRAB::T_fluid, 293.15, "Fluid temperature (K)");`

I provided the default to facilitate testing.  Is something wrong with this approach?

Thanks,
  Joe Kelly

**Comment by joe61vette** [comment, marked as answer]
Glad no one wasted their time on this.  It was my error.  I forgot to specify the temperature variable as a parameter to FluidMaterial.  One bad aspect of using a default rather than addRequiredCoupledVar.

At any rate, have a great weekend.

**Reply by GiudGiud** [reply]
Have a great weekend too!
Glad you got it solved
